pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -356,30 +360,47 @@ func validatePipelineParameterVariables(tasks []PipelineTask, params []ParamSpec
 		}
 	}
 
-	return validatePipelineVariables(tasks, "params", parameterNames, arrayParameterNames)
+	return validatePipelineParametersVariables(tasks, "params", parameterNames, arrayParameterNames)
 }
 
-func validatePipelineVariables(tasks []PipelineTask, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
+func validatePipelineParametersVariables(tasks []PipelineTask, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
 	for _, task := range tasks {
 		for _, param := range task.Params {
 			if param.Value.Type == ParamTypeString {
-				if err := validatePipelineVariable(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, paramNames); err != nil {
-					return err
-				}
-				if err := validatePipelineNoArrayReferenced(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, arrayParamNames); err != nil {
+				if err := validatePipelineStringVariable(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, paramNames, arrayParamNames); err != nil {
 					return err
 				}
 			} else {
 				for _, arrayElement := range param.Value.ArrayVal {
-					if err := validatePipelineVariable(fmt.Sprintf("param[%s]", param.Name), arrayElement, prefix, paramNames); err != nil {
-						return err
-					}
-					if err := validatePipelineArraysIsolated(fmt.Sprintf("param[%s]", param.Name), arrayElement, prefix, arrayParamNames); err != nil {
+					if err := validatePipelineArrayVariable(fmt.Sprintf("param[%s]", param.Name), arrayElement, prefix, paramNames, arrayParamNames); err != nil {
 						return err
 					}
 				}
 			}
 		}
+		if err := task.WhenExpressions.validateParametersVariables(prefix, paramNames, arrayParamNames); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validatePipelineStringVariable(name, value, prefix string, stringVars sets.String, arrayVars sets.String) *apis.FieldError {
+	if err := validatePipelineVariable(name, value, prefix, stringVars); err != nil {
+		return err
+	}
+	if err := validatePipelineNoArrayReferenced(name, value, prefix, arrayVars); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validatePipelineArrayVariable(name, value, prefix string, stringVars sets.String, arrayVars sets.String) *apis.FieldError {
+	if err := validatePipelineVariable(name, value, prefix, stringVars); err != nil {
+		return err
+	}
+	if err := validatePipelineArraysIsolated(name, value, prefix, arrayVars); err != nil {
+		return err
```

Comment:
ooo are these refactors that are unrelated to your changes? if so i highly recommend submitting them in a separate PR

Reply:
they are related because I use these functions to validate parameters in when expressions